channels/cliprdr/client/cliprdr_format.c

```diff
@@ -336,3 +336,181 @@ UINT cliprdr_process_format_data_response(cliprdrPlugin* cliprdr, wStream* s, UI
+		Stream_Seek(s, 16); /* reserved2 (16 bytes) */
+		Stream_Read_UINT64(s, lastWriteTime); /* lastWriteTime (8 bytes) */
+		file->ftLastWriteTime = uint64_to_filetime(lastWriteTime);
+		Stream_Read_UINT32(s, file->nFileSizeHigh); /* fileSizeHigh (4 bytes) */
+		Stream_Read_UINT32(s, file->nFileSizeLow); /* fileSizeLow (4 bytes) */
+		for (c = 0; c < 260; c++) /* cFileName (520 bytes) */
+			Stream_Read_UINT16(s, file->cFileName[c]);
+	}
+
+	if (Stream_GetRemainingLength(s) > 0)
+		WLog_WARN(TAG, "packed file list has %"PRIuz" excess bytes",
+			Stream_GetRemainingLength(s));
+out:
+	Stream_Free(s, FALSE);
+
+	return result;
+}
+
+#define CLIPRDR_MAX_FILE_SIZE (2U * 1024 * 1024 * 1024)
+
+/**
+ * Serialize a packed file list.
+ *
+ * The resulting format data must be freed with the `free()` function.
+ *
+ * @param [in]  file_descriptor_array  array of `FILEDESCRIPTOR` structs to serialize.
+ * @param [in]  file_descriptor_count  number of elements in `file_descriptor_array`.
+ * @param [out] format_data            serialized CLIPRDR_FILELIST.
+ * @param [out] format_data_length     length of `format_data` in bytes.
+ *
+ * @returns 0 on success, otherwise a Win32 error code.
+ */
+UINT cliprdr_serialize_file_list(const FILEDESCRIPTOR* file_descriptor_array,
+		UINT32 file_descriptor_count, BYTE** format_data, UINT32* format_data_length)
+{
+	UINT result = NO_ERROR;
+	UINT32 i;
+	wStream* s = NULL;
+
+	if (!file_descriptor_array || !format_data || !format_data_length)
+		return ERROR_BAD_ARGUMENTS;
+
+	s = Stream_New(NULL, 4 + file_descriptor_count * CLIPRDR_FILEDESCRIPTOR_SIZE);
+	if (!s)
+		return ERROR_NOT_ENOUGH_MEMORY;
+
+	Stream_Write_UINT32(s, file_descriptor_count); /* cItems (4 bytes) */
+
+	for (i = 0; i < file_descriptor_count; i++)
+	{
+		int c;
+		UINT64 lastWriteTime;
+		const FILEDESCRIPTOR* file = &file_descriptor_array[i];
+
+		/*
+		 * There is a known issue with Windows server getting stuck in
+		 * an infinite loop when downloading files that are larger than
+		 * 2 gigabytes. Do not allow clients to send such file lists.
+		 *
+		 * https://support.microsoft.com/en-us/help/2258090
+		 */
+		if ((file->nFileSizeHigh > 0) || (file->nFileSizeLow >= CLIPRDR_MAX_FILE_SIZE))
+		{
+			WLog_ERR(TAG, "cliprdr does not support files over 2 GB");
+			result = ERROR_FILE_TOO_LARGE;
+			goto error;
+		}
+
+		Stream_Write_UINT32(s, file->dwFlags); /* flags (4 bytes) */
+		Stream_Zero(s, 32); /* reserved1 (32 bytes) */
+		Stream_Write_UINT32(s, file->dwFileAttributes); /* fileAttributes (4 bytes) */
+		Stream_Zero(s, 16); /* reserved2 (16 bytes) */
+		lastWriteTime = filetime_to_uint64(file->ftLastWriteTime);
+		Stream_Write_UINT64(s, lastWriteTime); /* lastWriteTime (8 bytes) */
+		Stream_Write_UINT32(s, file->nFileSizeHigh); /* fileSizeHigh (4 bytes) */
+		Stream_Write_UINT32(s, file->nFileSizeLow); /* fileSizeLow (4 bytes) */
+		for (c = 0; c < 260; c++) /* cFileName (520 bytes) */
```

Comment:
@hardening, that functions are doing quite well for UTF-8—UTF-16 conversions, but here we want to convert between UTF-16 bytes and WCHARs, keeping the WCHARs because that's the type of the cFileName field of the FILEDESCRIPTOR struct.

The file name is Windows file name so Windows client using this clipboard function can directly use the FILEDESCRIPTOR struct without any additional work, while UNIX clients will have to convert UTF-16 into UTF-8 *and then* also replace backslashes with slashes. ConvertFromUnicode() won't do this for the client.

On Windows we use native FILEDESCRIPTOR definition. On non-Windows platforms we use our own struct from WinPR which is kept identical to the one on Windows. While it is totally possible to use a custom struct like WINPR_FILEDESCRIPTOR in both cases, I'd like to keep it named FILEDESCRIPTOR as in the spec and have it accurately represent the data received or sent to the server (which means UTF-16 file names with Windows-style separators). Non-Windows clients will not be able to use FILEDESCRIPTOR directly anyway, they will have to have their own custom structs for representing the files on the clipboard. Thus they can do any file name conversions they need by themselves.

Thus I do not see any problem with doing Stream_Read_UINT16() in a loop as that's exactly what we do here. @akallabeth's PR makes it prettier and more self-explanatory, so it would be nice to use those functions.